timemachine/cpp/src/segmented_weighted_random_sampler.cu

```diff
@@ -0,0 +1,128 @@
+#include "device_buffer.hpp"
+#include "gpu_utils.cuh"
+#include "kernels/k_sampling.cuh"
+#include "kernels/kernel_utils.cuh"
+#include "math_utils.cuh"
+#include "segmented_weighted_random_sampler.hpp"
+
+namespace timemachine {
+
+template <typename RealType>
+SegmentedWeightedRandomSampler<RealType>::SegmentedWeightedRandomSampler(
+    const int max_vals_per_segment, const int num_segments, const int seed)
+    : max_vals_per_segment_(max_vals_per_segment), num_segments_(num_segments), temp_storage_bytes_(0),
+      d_gumbel_(round_up_even(max_vals_per_segment_ * num_segments_)), d_arg_max_(num_segments_), d_argmax_storage_(0) {
+
+    int *dummy_segments = nullptr;
+    gpuErrchk(cub::DeviceSegmentedReduce::ArgMax(
+        d_argmax_storage_.data,
+        temp_storage_bytes_,
+        d_gumbel_.data,
+        d_arg_max_.data,
+        num_segments_,
+        dummy_segments,
+        dummy_segments + 1));
+    d_argmax_storage_.realloc(temp_storage_bytes_);
+
+    curandErrchk(curandCreateGenerator(&cr_rng_, CURAND_RNG_PSEUDO_DEFAULT));
+    curandErrchk(curandSetPseudoRandomGeneratorSeed(cr_rng_, seed));
+};
+
+template <typename RealType> SegmentedWeightedRandomSampler<RealType>::~SegmentedWeightedRandomSampler() {
+    curandErrchk(curandDestroyGenerator(cr_rng_));
+};
+
+template <typename RealType>
+void SegmentedWeightedRandomSampler<RealType>::sample_device(
+    const int total_values,
+    const int num_segments,
+    const int *d_segments,               // [num_segments + 1]
+    const RealType *d_log_probabilities, // [total_values]
+    int *d_samples,                      // [num_segments]
+    cudaStream_t stream) {
+    // Perform sampling using the gumbel-max-trick like cupy
+    // https://github.com/cupy/cupy/blob/46f6ed6e78661fb1d31b41a072519f119f5a2385/cupy/random/_generator.py#L1115-L1121
+    curandErrchk(curandSetStream(cr_rng_, stream));
+    curandErrchk(templateCurandUniform(cr_rng_, d_gumbel_.data, d_gumbel_.length));
+
+    // Use the noise both as the noise and the intermediate, does change the values in d_gumbel
```

Comment:
Without knowing the details of the trick used here, this sounds a bit suspicious. Are we sure this is valid and won't bias the sampling in some way?

Reply:
If the sampling was biased I would expect this to show up in the tests. This is done because you need to transform the noise before doing argmax (https://github.com/proteneer/timemachine/blob/master/timemachine/cpp/src/kernels/k_sampling.cuh#L20-L22) the only reason we don't want to modify it in practice is because we want to re-use the same noise repeatedly times to get the determinism between batched and serial versions.

Reply:
Got it, I think that makes sense. The comment initially made me think that the same random values were being used twice, but it sounds like we had just been reusing the buffer and overwriting the initial values (and since then have changed to not overwrite, so the point is moot)?